# Task 4: Satisfaction Analysis

**Loading the Processed Data saved as csv File in Task 3**

In [3]:
# Step 1: Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# Step 2: Loading the Data saved in Task 2
df = pd.read_csv(r'processed_data.csv')

# Show all columns
pd.set_option('display.max_columns', None)
# Set to allow auto-fit
pd.set_option('display.max_colwidth', None)
# Prevent scientific notation
pd.set_option('display.float_format', '{:.0f}'.format)

# Displaying first few rows
df.head()

Bearer Id                Start  Start ms                  End  \
0  13114483460844900352  2019-04-04 12:01:18       770  2019-04-25 14:35:31   
1  13114483482878900224  2019-04-09 13:04:04       235  2019-04-25 08:15:48   
2  13114483484080500736  2019-04-09 17:42:11         1  2019-04-25 11:58:13   
3  13114483485442799616  2019-04-10 00:31:25       486  2019-04-25 07:36:35   
4  13114483499480700928  2019-04-12 20:10:23       565  2019-04-25 10:40:32   

   End ms  Dur. (ms)            IMSI  MSISDN/Number           IMEI  \
0     662    1823652 208201448079117    33664962239 35521209507511   
1     606    1365104 208201909211140    33681854413 35794009006359   
2     652    1361762 208200314458056    33760627129 35281510359387   
3     171    1321509 208201402342131    33750343200 35356610164913   
4     954    1089009 208201401415120    33699795932 35407009745539   

  Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0   9164566995485190               42                5   
1            L77566A               65                5   
2            D42335A              110               18   
3            T21824A              110               18   
4            D88865A              110               18   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                       23                       44   
1                       16                       26   
2                        6                        9   
3                       44                       44   
4                        6                        9   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                     20809914                       759659   
1                     20809914                       759659   
2                     20809914                       759659   
3                     20809914                       759659   
4                     20809914                       759659   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                  100                               0   
1                  100                               0   
2                  100                               0   
3                  100                               0   
4                  100                               0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                              0                   0                  100   
1                              0                   0                  100   
2                              0                   0                  100   
3                              0                   0                  100   
4                              0                   0                  100   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                              0                               0   
1                              0                               0   
2                              0                               0   
3                              0                               0   
4                              0                               0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                     0        114471024          3242301   
1                     0        114471024          3242301   
2                     0        114471024          3242301   
3                     0        114471024          3242301   
4                     0        114471024          3242301   

   Activity Duration DL (ms)  Activity Duration UL (ms)  Dur. (ms).1  \
0                      37624                      38787   1823652892   
1                        168                       3560   1365104371   
2                          0                          0   1361762651   
3                       3330                      37882   1321509685   
4                          0                          0   1089009389   

  Handset Manufacturer                

In [5]:
# Step 3: Loading Aggregated DataFrame saved in Task 2
aggregated = pd.read_csv(r'aggregated_results.csv')

# Show all columns
pd.set_option('display.max_columns', None)
# Set to allow auto-fit
pd.set_option('display.max_colwidth', None)
# Prevent scientific notation
pd.set_option('display.float_format', '{:.0f}'.format)

# Displaying first 10 rows
aggregated.head(10)

MSISDN/Number  avg_tcp_retransmission  avg_rtt  \
0    33601001722                21569573       46   
1    33601001754                21569573       31   
2    33601002511                21569573      127   
3    33601007832                  760725       84   
4    33601008617                15470202       60   
5    33601010682                11165996       76   
6    33601011634                10839902       26   
7    33601011959                  759937       52   
8    33601014694                21569573      127   
9    33601020306                20811208       62   

                                       handset_type  avg_throughput  \
0                    Huawei P20 Lite Huawei Nova 3E              76   
1                            Apple iPhone 7 (A1778)              99   
2                                         undefined              97   
3                           Apple iPhone 5S (A1457)             248   
4                           Apple iPhone Se (A1723)           26152   
5                          Samsung Galaxy A8 (2018)            3954   
6  Huawei Mate 10 Pro Porsche Design Huawei Mate 10           21256   
7                 Samsung Galaxy S8 Plus (Sm-G955F)            1247   
8                                         undefined              94   
9                            Apple iPhone X (A1865)             146   

   experience_cluster  
0                   0  
1                   0  
2                   0  
3                   1  
4                   1  
5                   0  
6                   1  
7                   1  
8                   0  
9                   0

In [6]:
# Step 4: Loading Engagement DataFrame saved in Task 2
engagement = pd.read_csv(r'user_engagement_data.csv')

# Show all columns
pd.set_option('display.max_columns', None)
# Set to allow auto-fit
pd.set_option('display.max_colwidth', None)

# Displaying first 10 rows
engagement.head(10)

MSISDN/Number  session_frequency  session_duration   total_dl  total_ul  \
0    33601001722                  1            116720  842637466  36053108   
1    33601001754                  1            181230  120755184  36104459   
2    33601002511                  1            134969  556659663  39306820   
3    33601007832                  1             49878  401993172  20327526   
4    33601008617                  2             37104 1363130417  94280527   
5    33601010682                  2            253983  555207972  60009249   
6    33601011634                  2            128360  553041208 101681858   
7    33601011959                  1             86399  294085078  38575279   
8    33601014694                  2            495702  896560173  93572016   
9    33601020306                  1            124854  685372753  47091008   

   total_social_media_dl  total_social_media_ul  total_youtube_dl  \
0                2206504                  25631          14900201   
1                2598548                  62017           5324251   
2                3148004                  47619           2137272   
3                 251469                  28825           4393123   
4                2861230                  51312          10339971   
5                3787122                   5790          12169742   
6                2707339                  62288           2990819   
7                1740000                  42692           5100103   
8                5206619                  82066          41189830   
9                 938582                  18193          13624376   

   total_youtube_ul  total_netflix_dl  total_netflix_ul  total_google_dl  \
0           6724347          10265105          16915876          3337123   
1           7107972            770569          10451194          4197697   
2          19196298          16525919           2827981          3343483   
3           2584198           1157362            784730          5937765   
4          31193031          24971647          24230077         13728668   
5          22177380          14569029          15746472         11993496   
6          23440516          12813052          32999345          4346088   
7          13792637           1392985          13899855         11036391   
8          22253972          27418209          26797908         11046222   
9          14945336          22460852          13064219         11098762   

   total_google_ul  total_email_dl  total_email_ul  total_gaming_dl  \
0          1051882          837400          493962        811091133   
1          1137166         2828821          478960        105035298   
2            99643         2436500          768880        529068485   
3          3740728         2178618          106052        388074835   
4          4770948         2247808         1057661       1308981093   
5          3269638         6513834         1334762        506174749   
6          5850315          736446          622513        529447464   
7           987223         1280558          935005        273535041   
8          3087989         3666031          933794        808033262   
9          1836407         1986925          796355        635263256   

   total_gaming_ul  total_other_dl  total_other_ul  total_traffic  \
0          1367528       377096990         9473882      878690574   
1         14714780       279557701         2152370      156859643   
2          9759228       495086501         6607171      595966483   
3          3051292        25248001        10031701      422320698   
4          5816727       777643713        27160771     1457410944   
5          5461038       480358608        12014169      615217221   
6         19013593       205195425        19693288      654723066   
7          8829356       178113026           88511      332660357   
8         25054690      1026521395        15361597      990132189   
9         11832418       133741168         4598080      732463761   

   to

### Summary of Findings

1. **Data Loading**:
   - **Loaded Datasets**: 
     - Used `pd.read_csv()` to load three datasets:
       - `processed_data.csv`
       - `aggregated_results.csv`
       - `user_engagement_data.csv`

2. **Display Settings**:
   - **Configured Display Options**:
     - Set `pd.set_option('display.max_columns', None)` to show all columns.
     - Set `pd.set_option('display.max_colwidth', None)` to allow auto-fit for better readability.
     - Set `pd.set_option('display.float_format', '{:.0f}'.format)` to prevent scientific notation in numerical data.

**Task 4.1 - Assign Scores**

In [8]:
import pandas as pd

# Assuming `df` contains engagement-related data
df_engagement = engagement[[
    'MSISDN/Number',
    'session_frequency',
    'session_duration',
    'total_traffic','cluster'
]].copy()

# Assuming `aggregated` contains experience-related data
df_experience = aggregated[[
    'MSISDN/Number',
    'avg_tcp_retransmission',
    'avg_rtt',
    'avg_throughput', 'experience_cluster'
]].copy()

# Show all columns
pd.set_option('display.width', 1000)

# Display the extracted DataFrames
print("Engagement DataFrame:")
print(df_engagement.head())

print("\nExperience DataFrame:")
print(df_experience.head())

Engagement DataFrame:
   MSISDN/Number  session_frequency  session_duration  total_traffic  cluster
0    33601001722                  1            116720      878690574        0
1    33601001754                  1            181230      156859643        0
2    33601002511                  1            134969      595966483        0
3    33601007832                  1             49878      422320698        0
4    33601008617                  2             37104     1457410944        0

Experience DataFrame:
   MSISDN/Number  avg_tcp_retransmission  avg_rtt  avg_throughput  experience_cluster
0    33601001722                21569573       46              76                   0
1    33601001754                21569573       31              99                   0
2    33601002511                21569573      127              97                   0
3    33601007832                  760725       84             248                   1
4    33601008617                15470202       60        

In [9]:
import numpy as np
import pandas as pd

def assign_scores(df_engagement, df_experience):
    """Assigns engagement and experience scores."""

    # Get the cluster ID of the cluster with the min average session duration and min total traffic
    min_engagement_cluster = df_engagement.groupby('cluster')[['session_duration', 'total_traffic']].mean().sum(axis=1).idxmin()

    # Get the cluster ID of the cluster with the highest average TCP retransmission and highest average RTT
    max_experience_cluster = df_experience.groupby('experience_cluster')[['avg_tcp_retransmission', 'avg_rtt']].mean().sum(axis=1).idxmax()

    # Calculate cluster centers once
    engagement_centers = df_engagement.groupby('cluster')[['session_frequency', 'session_duration', 'total_traffic']].mean()
    experience_centers = df_experience.groupby('experience_cluster')[['avg_tcp_retransmission', 'avg_rtt', 'avg_throughput']].mean()

    # Assign engagement scores
    def assign_engagement_score(row):
        return np.linalg.norm(row[['session_frequency', 'session_duration', 'total_traffic']].values - engagement_centers.loc[min_engagement_cluster].values)

    df_engagement['engagement_score'] = df_engagement.apply(assign_engagement_score, axis=1)

    # Assign experience scores
    def assign_experience_score(row):
        return np.linalg.norm(row[['avg_tcp_retransmission', 'avg_rtt', 'avg_throughput']].values - experience_centers.loc[max_experience_cluster].values)

    df_experience['experience_score'] = df_experience.apply(assign_experience_score, axis=1)

    # Merge engagement and experience score into the same data frame by MSISDN/Number
    df_engagement_scores = df_engagement[['MSISDN/Number', 'engagement_score']].copy()
    df_experience_scores = df_experience[['MSISDN/Number', 'experience_score']].copy()

    scores_df = pd.merge(df_engagement_scores, df_experience_scores, on='MSISDN/Number', how='inner')
    return scores_df

# Call the function
try:
    scores_df = assign_scores(df_engagement, df_experience)
    print(scores_df)
except Exception as e:
    print(f"An error occurred: {e}")

         MSISDN/Number  engagement_score  experience_score
0          33601001722         203595411            523472
1          33601001754         518235523            523472
2          33601002511          79128681            523472
3          33601007832         252774480          21332315
4          33601008617         782315786           6622880
...                ...               ...               ...
106852     33789997247         194787829            523472
106853     41882819545      531069619018          12077814
106854   3197020876596         442971815            523473
106855 337000037000919          78807429            523473
106856 882397108489451         535942022            523473

[106857 rows x 3 columns]


### Summary of Findings

1. **Data Preparation**:
   - **Created `df_engagement`**: Extracted relevant columns from the engagement data.
   - **Created `df_experience`**: Extracted relevant columns from the experience data.

2. **Display Settings**:
   - Set display width to 1000 for better visibility of DataFrame columns.

3. **Function Definition**:
   - **Defined `assign_scores` Function**:
     - **Identified Clusters**: Found the cluster with the lowest average session duration and traffic for engagement, and the cluster with the highest TCP retransmission and RTT for experience.
     - **Calculated Averages**: Computed average metrics for each cluster.
     - **Assigned Scores**: Used Euclidean distance to calculate engagement and experience scores.
     - **Merged Scores**: Combined scores into a single DataFrame based on `MSISDN/Number`.

4. **Function Execution**:
   - **Executed the Function**: Called `assign_scores` and printed the resulting DataFrame, handling any errors that occurred.

### Output
- The result is a DataFrame (`scores_df`) containing `MSISDN/Number`, `engagement_score`, and `experience_score`. 

### Error Handling
- Included error handling to catch and print any issues during execution.

**Task 4.2 - Satisfaction Score**

In [11]:
import numpy as np
import pandas as pd

def calculate_satisfaction_score(df):
    """Calculates the average satisfaction score and identifies satisfied customers."""
    
    # Calculate the satisfaction score
    df['satisfaction_score'] = (df['engagement_score'] + df['experience_score']) / 2
    
    # Get the top 10 satisfied customers
    top_10_satisfied = df.nlargest(10, 'satisfaction_score')
    
    # Print the top 10 satisfied customers
    print("Top 10 Satisfied Customers:\n", top_10_satisfied[['MSISDN/Number', 'satisfaction_score']])
    
    return df

updated_df = calculate_satisfaction_score(scores_df)  # Call the function with the scores DataFrame
# Print the updated DataFrame
print("\nUpdated DataFrame:\n",updated_df.head())

Top 10 Satisfied Customers:
         MSISDN/Number  satisfaction_score
106853    41882819545        265540848416
6437      33614892860          4092661688
92923     33760536639          3922548365
13180     33625779332          3915608881
13526     33626320676          3652505970
76363     33675877202          3613476583
37052     33659725664          3523629391
63028     33666464084          3321795709
92577     33760413819          3234885129
57241     33664712899          3104326093

Updated DataFrame:
    MSISDN/Number  engagement_score  experience_score  satisfaction_score
0    33601001722         203595411            523472           102059442
1    33601001754         518235523            523472           259379498
2    33601002511          79128681            523472            39826077
3    33601007832         252774480          21332315           137053398
4    33601008617         782315786           6622880           394469333


In [ ]:
### Summary of Actions Taken

1. **Function Definition**:
   - **`calculate_satisfaction_score`**:
     - **Calculated Satisfaction Score**: Averaged `engagement_score` and `experience_score`.
     - **Top Customers**: Identified the top 10 customers with the highest satisfaction scores.

2. **Function Execution**:
   - Called the function with `scores_df`.
   - Printed the top 10 satisfied customers and the updated DataFrame.

### Output

- **Top 10 Satisfied Customers**:
  ```
      MSISDN/Number  satisfaction_score
  106853    41882819545        265540848416
  6437      33614892860          4092661688
  92923     33760536639          3922548365
  13180     33625779332          3915608881
  13526     33626320676          3652505970
  ```

- **Updated DataFrame**:
  ```
      MSISDN/Number  engagement_score  experience_score  satisfaction_score
  0    33601001722         203595411            523472           102059442
  1    33601001754         518235523            523472           259379498
  ```

### Conclusion

- The function successfully calculated satisfaction scores and highlighted the top customers.

## Summary of Findings

I created a function to calculate customer satisfaction scores by averaging their engagement and experience scores. The process involved:

1. **Calculating Satisfaction Score**: Added a new column, `satisfaction_score`, using the formula:
   satisfaction_score = (engagement_score + experience_score) / 2

2. **Identifying Top Customers**: Extracted the top 10 customers with the highest satisfaction scores.
## Outcome


The analysis identified the top 10 satisfied customers and updated the DataFrame to include their satisfaction scores, enabling further insights into customer satisfaction.

**Task 4.3 - Predicting Satisfaction**

In [13]:
# Check the columns in the DataFrame
print("Columns in clustered_df:", updated_df.columns)

# Strip any leading or trailing whitespace from column names
updated_df.columns = updated_df.columns.str.strip()

# Re-check the columns after stripping
print("Columns after stripping:", updated_df.columns)

Columns in clustered_df: Index(['MSISDN/Number', 'engagement_score', 'experience_score', 'satisfaction_score'], dtype='object')
Columns after stripping: Index(['MSISDN/Number', 'engagement_score', 'experience_score', 'satisfaction_score'], dtype='object')


In [14]:
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor

# Define features and target
X = updated_df[['engagement_score', 'experience_score']]
y = updated_df['satisfaction_score']

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and fit the model
model = RandomForestRegressor()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"\nModel RMSE: {rmse:.2f}")

# Cross-validation scores
cv_scores = cross_val_score(model, X, y, cv=5, scoring='neg_root_mean_squared_error')
print(f"\nCross-validated RMSE: {-cv_scores.mean():.2f}")

# Inspect predictions
comparison_df = pd.DataFrame({
    'Actual': y_test,
    'Predicted': y_pred,
    'Difference': y_test - y_pred,
    'Absolute Difference': abs(y_test - y_pred)
})
print(comparison_df)


Model RMSE: 758662.03

Cross-validated RMSE: 359021237.21
         Actual  Predicted  Difference  Absolute Difference
21377  65473889   65522767      -48878                48878
22356 280635546  280609343       26203                26203
42414  19178307   19091523       86784                86784
95160 435144978  434311450      833528               833528
61074  40292442   40272947       19494                19494
...         ...        ...         ...                  ...
57023 279638926  279820399     -181474               181474
39594 279988165  279988894        -728                  728
36906 279070593  279071959       -1367                 1367
16701 627913549  628337911     -424361               424361
92311  24415385   24416816       -1431                 1431

[21372 rows x 4 columns]


In [15]:
from sklearn.metrics import r2_score

# Calculate R² score
r2 = r2_score(y_test, y_pred)
print(f"R² Score: {r2:.2f}")

R² Score: 1.00


**Saving the Model to a Pickle File for Future use**

In [17]:
import pickle

# Save the model to a pickle file
with open('random_forest_model.pkl', 'wb') as file:
    pickle.dump(model, file)

print("Model saved to random_forest_model.pkl")

Model saved to random_forest_model.pkl


## Summary of Findings

In this analysis, I built a Random Forest regression model to predict customer satisfaction scores based on engagement and experience scores. Here’s a concise overview of the process:

### Process
1. **Data Preparation**:
   - Checked and stripped whitespace from the DataFrame column names.

2. **Feature and Target Definition**:
   - Defined features (`X`) as `engagement_score` and `experience_score`.
   - Defined the target variable (`y`) as `satisfaction_score`.

3. **Data Splitting**:
   - Split the data into training and testing sets (80% train, 20% test).

4. **Model Creation and Training**:
   - Created a Random Forest Regressor model and fitted it to the training data.

5. **Predictions**:
   - Made predictions on the test set.

6. **Model Evaluation**:
   - Calculated Root Mean Squared Error (RMSE) and cross-validated RMSE.
   - Generated a comparison DataFrame of actual vs. predicted values.

7. **R² Score Calculation**:
   - Calculated the R² score to assess the model's performance.

8. **Model Saving**:
   - Saved the trained model to a pickle file for future use.

### Outcome
- **Model RMSE**: 758,662.03
- **Cross-Validated RMSE**: 359,021,237.21
- **R² Score**: 1.00 (indicating perfect fit)

The model was successfully saved to `random_forest_model.pkl`, allowing for future predictions without retraining.

**Task 4.4 - Clustering Satisfaction and Experience**

In [19]:
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

def cluster_satisfaction_experience(df):
    """Clusters users based on satisfaction and experience scores."""
    
    # Clustering based on k-means using engagement and experience
    scaler = StandardScaler()
    scaled_scores = scaler.fit_transform(df[['engagement_score', 'experience_score']])
    
    kmeans = KMeans(n_clusters=2, random_state=42)
    df['satisfaction_cluster'] = kmeans.fit_predict(scaled_scores)
    
    # Print the first 20 cluster values
    print(f"\nCluster values:\n {df[['MSISDN/Number', 'satisfaction_cluster']].head(20)}")
    
    return df

# Now call the clustering function
clustered_df = cluster_satisfaction_experience(updated_df)


Cluster values:
     MSISDN/Number  satisfaction_cluster
0     33601001722                     0
1     33601001754                     0
2     33601002511                     0
3     33601007832                     0
4     33601008617                     0
5     33601010682                     0
6     33601011634                     0
7     33601011959                     0
8     33601014694                     0
9     33601020306                     0
10    33601021045                     0
11    33601021217                     0
12    33601022743                     0
13    33601024291                     0
14    33601025738                     0
15    33601026147                     0
16    33601027208                     0
17    33601031129                     0
18    33601032846                     0
19    33601032987                     0


In [20]:
print(updated_df.head())

   MSISDN/Number  engagement_score  experience_score  satisfaction_score  satisfaction_cluster
0    33601001722         203595411            523472           102059442                     0
1    33601001754         518235523            523472           259379498                     0
2    33601002511          79128681            523472            39826077                     0
3    33601007832         252774480          21332315           137053398                     0
4    33601008617         782315786           6622880           394469333                     0


## Summary of Findings

In this analysis, I applied K-Means clustering to group users based on their satisfaction and experience scores.

### Process
1. **Function Creation**: Developed a function to perform clustering.
2. **Data Standardization**: Used `StandardScaler` to normalize the `engagement_score` and `experience_score` for better accuracy.
3. **K-Means Application**: Implemented K-Means with 2 clusters to categorize users.

### Outcome
- The clustering results showed minimal variation among the groups, suggesting that the features used may not effectively distinguish between users. This indicates a need for further refinement in the analysis.

**Task 4.5 - Aggregate Scores per Cluster**

In [22]:
def aggregate_scores_per_cluster(df):
    """Aggregates average satisfaction and experience scores per cluster."""
    
    # Group by satisfaction cluster and calculate mean scores
    cluster_scores = df.groupby('satisfaction_cluster')[['satisfaction_score', 'experience_score']].mean()
    
    # Print the average scores per cluster
    print("\nAverage satisfaction and experience scores per cluster:\n\n", cluster_scores)
    
    return cluster_scores

# Now call the aggregation function
average_scores = aggregate_scores_per_cluster(clustered_df)


Average satisfaction and experience scores per cluster:

                       satisfaction_score  experience_score
satisfaction_cluster                                      
0                              174476438           8365504
1                           265540848416          12077814


## Summary of Findings

In this analysis, I calculated the average satisfaction and experience scores for each user cluster.

### Process
1. **Function Development**: Created a function to aggregate scores by cluster.
2. **Aggregation**: Used `groupby` to calculate the mean satisfaction and experience scores for each cluster.

### Outcome
- The average scores revealed significant differences between the clusters, indicating that the users in different groups have varying levels of satisfaction and experience. This suggests that the clustering approach successfully identified distinct user segments.

**Task 4.6 - Export to MySQL**

In [24]:
!pip install sqlalchemy

In [25]:
!pip install pyodbc

In [28]:
from sqlalchemy import create_engine, text

# Define SQL Server connection details
server = 'DEB\\SQLEXPRESS'
database = 'Debais_Baidya'
username = 'Debasis'
password = '12345'

# Create a connection string
connection_url = f"mssql+pyodbc://{username}:{password}@{server}/{database}?driver=ODBC+Driver+17+for+SQL+Server"

# Create an engine
engine = create_engine(connection_url)

# Create Table
create_table_query = text("""
IF NOT EXISTS (SELECT * FROM sysobjects WHERE name='user_scores' AND xtype='U')
BEGIN
CREATE TABLE user_scores (
    [MSISDN/Number] VARCHAR(255) PRIMARY KEY,
    engagement_score FLOAT,
    experience_score FLOAT,
    satisfaction_score FLOAT,
    satisfaction_cluster INT
)
END
""")

# Assuming you have a connection object created from the engine
with engine.connect() as connection:
    connection.execute(create_table_query)

    # Prepare insert query
    insert_query = text("""
    MERGE INTO user_scores AS target
    USING (SELECT :msisdn AS [MSISDN/Number], :engagement AS engagement_score,
                  :experience AS experience_score, :satisfaction AS satisfaction_score,
                  :cluster AS satisfaction_cluster) AS source
    ON target.[MSISDN/Number] = source.[MSISDN/Number]
    WHEN MATCHED THEN
        UPDATE SET
            target.engagement_score = source.engagement_score,
            target.experience_score = source.experience_score,
            target.satisfaction_score = source.satisfaction_score,
            target.satisfaction_cluster = source.satisfaction_cluster
    WHEN NOT MATCHED THEN
        INSERT ([MSISDN/Number], engagement_score, experience_score, satisfaction_score, satisfaction_cluster)
        VALUES (source.[MSISDN/Number], source.engagement_score, source.experience_score, source.satisfaction_score, source.satisfaction_cluster);
    """)

    # Insert data using MERGE
    for _, row in updated_df.iterrows():
        # Execute the insert query with parameters as a dictionary
        connection.execute(insert_query, {
            'msisdn': str(row['MSISDN/Number']),  # Ensure it's a string
            'engagement': row['engagement_score'],
            'experience': row['experience_score'],
            'satisfaction': row['satisfaction_score'],
            'cluster': row['satisfaction_cluster']
        })

    connection.commit()  # Commit once after all inserts

print("Data exported to SQL Server successfully.")

# Dispose the engine
engine.dispose()

Data exported to SQL Server successfully.


**Query the Information Schema from Python directly to SQL SSMS**

In [30]:
# Check if the table exists
check_table_query = text("""
SELECT *
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_NAME = 'user_scores';
""")

with engine.connect() as connection:
    result = connection.execute(check_table_query)
    table_info = result.fetchall()

if table_info:
    print("Table 'user_scores' exists.")
else:
    print("Table 'user_scores' does not exist.")

Table 'user_scores' exists.


**Describe the Table Structure**

In [33]:
# Describe the table structure
describe_table_query = text("""
SELECT COLUMN_NAME, DATA_TYPE, CHARACTER_MAXIMUM_LENGTH
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'user_scores';
""")

with engine.connect() as connection:
    result = connection.execute(describe_table_query)
    columns_info = result.fetchall()

print("Table 'user_scores' structure:")
for column in columns_info:
    print(f"Column: {column.COLUMN_NAME}, Type: {column.DATA_TYPE}, Length: {column.CHARACTER_MAXIMUM_LENGTH}")

Table 'user_scores' structure:
Column: MSISDN/Number, Type: varchar, Length: 255
Column: engagement_score, Type: float, Length: None
Column: experience_score, Type: float, Length: None
Column: satisfaction_score, Type: float, Length: None
Column: satisfaction_cluster, Type: int, Length: None


## Summary of Database Operations

In this section, I detailed the process of creating and populating the `user_scores` table in SQL Server.

### Table Structure
- **MSISDN/Number**: `varchar(255)` (Primary Key)
- **engagement_score**: `float`
- **experience_score**: `float`
- **satisfaction_score**: `float`
- **satisfaction_cluster**: `int`

### Process
1. **Install Required Libraries**: Utilized `sqlalchemy` and `pyodbc` for database connectivity.
2. **Connection Setup**: Defined connection details for SQL Server and created a connection string.
3. **Table Creation**: Executed a query to create the `user_scores` table if it did not already exist.
4. **Data Insertion**: Used a `MERGE` statement to insert or update user scores based on the existing records in the table.

### Outcome
- Successfully exported data to SQL Server. 
- Verified the existence of the `user_scores` table and described its structure, confirming that it matches the intended design. 

### Verification
- **Table Existence**: Confirmed that the `user_scores` table exists.
- **Table Structure**: Retrieved and displayed the column names, data types, and lengths, ensuring the table structure is as expected.